src/tasks/getsubscriptiondetails/taskgetsubscriptiondetails.cpp

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "taskgetsubscriptiondetails.h"
+#include "authenticationlistener.h"
+#include "authenticationinapp/authenticationinappsession.h"
+#include "errorhandler.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+
+namespace {
+Logger logger(LOG_MAIN, "TaskGetSubscriptionDetails");
+}  // anonymous namespace
+
+TaskGetSubscriptionDetails::TaskGetSubscriptionDetails(
+    const QString& emailAddress)
+    : Task("TaskGetSubscriptionDetails"), m_emailAddress(emailAddress) {
+  MVPN_COUNT_CTOR(TaskGetSubscriptionDetails);
+}
+
+TaskGetSubscriptionDetails::~TaskGetSubscriptionDetails() {
+  MVPN_COUNT_DTOR(TaskGetSubscriptionDetails);
+}
+
+void TaskGetSubscriptionDetails::run() {
+  logger.debug() << "run";
+
+  NetworkRequest* request =
+      NetworkRequest::createForGetSubscriptionDetails(this);
+
+  connect(
+      request, &NetworkRequest::requestFailed, this,
+      [this](QNetworkReply::NetworkError error, const QByteArray&) {
+        logger.error() << "Get subscription details failed" << error;
+
+        // Network request failed after authentication for a second time
+        if (m_authenticationInAppSession) {
+          MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+          emit failed();
+          m_authenticationInAppSession->terminate();
+          return;
+        }
+
+        // User needs to (re)authenticate
+        if (error == QNetworkReply::AuthenticationRequiredError) {
+          emit needsAuthentication();
+          initAuthentication();
+          return;
+        }
+
+        MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+        emit failed();
+        emit completed();
+      });
+
+  connect(request, &NetworkRequest::requestCompleted, this,
+          [this](const QByteArray& data) {
+            logger.debug() << "Get subscription details completed" << data;
+
+            emit receivedData(data);
+
+            if (m_authenticationInAppSession) {
+              m_authenticationInAppSession->terminate();
+              return;
+            }
```

Comment:
can you invert these 2 blocks? First `terminate` then `emit`?

Reply:
Currently, when terminating the auth the `Task` will be completed [here](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/3606/files/bbaddfab08d031a7b73226fd9a25390be4f1b941#diff-0de758d86f6de3f1fc61b369c831dfc6c1f3e4458dff91ab3a4312f604b2f8c6R84-R85). Should I change that behaviour?